server/src/main/java/org/elasticsearch/cluster/routing/IndexRoutingTable.java

```diff
@@ -481,29 +481,56 @@ private Builder initializeEmpty(IndexMetadata indexMetadata, UnassignedInfo unas
                         )
                     );
                 }
-                shards.put(shardNumber, indexShardRoutingBuilder.build());
+                shards[shardNumber] = indexShardRoutingBuilder;
             }
             return this;
         }
 
         public Builder addReplica() {
-            for (var shardNumber : shards.keySet()) {
-                ShardId shardId = new ShardId(index, shardNumber);
-                // version 0, will get updated when reroute will happen
-                ShardRouting shard = ShardRouting.newUnassigned(
-                    shardId,
-                    false,
-                    PeerRecoverySource.INSTANCE,
-                    new UnassignedInfo(UnassignedInfo.Reason.REPLICA_ADDED, null)
-                );
-                shards.put(shardNumber, new IndexShardRoutingTable.Builder(shards.get(shard.id())).addShard(shard).build());
+            if (shards == null) {
+                return this;
+            }
+            for (int shardNumber = 0; shardNumber < shards.length; shardNumber++) {
+                Object existing = shards[shardNumber];
+                if (existing == null) {
+                    continue;
+                }
+                if (existing instanceof IndexShardRoutingTable) {
+                    shards[shardNumber] = new IndexShardRoutingTable.Builder((IndexShardRoutingTable) existing).addShard(
+                        newUnassignedReplica(((IndexShardRoutingTable) existing).shardId())
+                    ).build();
+                } else {
+                    assert existing instanceof IndexShardRoutingTable.Builder;
+                    // version 0, will get updated when reroute will happen
```

Comment:
I'd think this comment also applies in the case of `IndexShardRoutingTable`?

Reply:
Only the builder version is left :) -> problem solved